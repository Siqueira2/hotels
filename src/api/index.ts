export * from './places'
export * from './hotels'
