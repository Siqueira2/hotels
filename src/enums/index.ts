export enum AmenitiesEnum {
  wifi = 'WI_FI',
  restaurant = 'RESTAURANT',
  room_service = 'ROOM_SERVICE',
  meeting_room = 'MEETING_ROOM',
  local_laundry_service = 'LAUNDRY',
  bakery_dining = 'BREAKFAST',
  local_parking = 'PARKING',
  fitness_center = 'FITNESS_CENTER',
  pool = 'POOL',
  self_improvement = 'SPA',
  hot_tub = 'STEAM_ROOM',
  pets = 'PETS',
  sports_bar = 'PUB',
  ac_unit = 'AIR_CONDITIONING',
  lock = 'SAFE',
  tv = 'ROOM_TV',
  timer = 'RECEPTION_24_HOURS',
  vertical_shades = 'STAGE',
  vertical_shades_closed = 'AUDITORIUM',
  spa = 'MASSAGE',
  mood = 'PLAY_GROUND',
  cases = 'TRANSFER',
}
