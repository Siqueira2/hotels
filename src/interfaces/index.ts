export interface State {
  name: string
  shortname: string
}

export interface Place {
  state: State
  name: string
  placeId: number
}
