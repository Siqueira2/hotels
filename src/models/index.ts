export * from './place'
export * from './hotel'
export * from './placeHotel'
export * from './hotelFilterService'
