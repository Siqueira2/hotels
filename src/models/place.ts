import type { Place as PlaceType, State } from 'src/interfaces'

export class Place {
  private placeId: number
  private name: string
  private state: State

  constructor(place: PlaceType) {
    this.placeId = place.placeId
    this.name = place.name
    this.state = place.state
  }

  public get getPlaceId(): number {
    return this.placeId
  }

  public get getName(): string {
    return this.name
  }

  public get getState(): State {
    return this.state
  }

  public get formattedName(): string {
    return `${this.name}, ${this.state.name}`
  }

  public get formattedNameWithShortName(): string {
    const words = this.name.split(' ')
    let initials = ''

    if (words.length === 1) {
      initials = words[0]?.slice(0, 3).toUpperCase() || ''
    }

    if (words.length > 2) {
      const firstInitial = words[0]?.charAt(0).toUpperCase() || ''
      const lastInitial = words[words.length - 1]?.charAt(0).toUpperCase() || ''
      initials = `${firstInitial}${lastInitial}`
    }

    initials ||= words.map((word) => word?.charAt(0).toUpperCase() || '').join('')

    return `${this.name}, ${initials}`
  }
}
