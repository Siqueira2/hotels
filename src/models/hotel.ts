import type { Hotel as HotelType, Address, Amenity } from 'src/interfaces'

import { formatCurrency } from 'src/helpers'

export class Hotel {
  private id: number
  private name: string
  private description: string
  private stars: string
  private thumb: string
  private favorite: boolean
  private price: number
  private address: Address
  private amenities: Amenity[] | null
  private hasBreakFast: boolean
  private hasRefundableRoom: boolean
  private hasAgreement: boolean
  private nonRefundable: boolean | null
  private roomsQuantity: number
  private images: string[]

  constructor(hotel: HotelType) {
    this.id = hotel.id
    this.name = hotel.name
    this.description = hotel.description
    this.stars = hotel.stars
    this.thumb = hotel.thumb
    this.favorite = hotel.favorite
    this.price = hotel.price
    this.address = hotel.address
    this.amenities = hotel.amenities
    this.hasBreakFast = hotel.hasBreakFast
    this.hasRefundableRoom = hotel.hasRefundableRoom
    this.hasAgreement = hotel.hasAgreement
    this.nonRefundable = hotel.nonRefundable
    this.roomsQuantity = hotel.roomsQuantity
    this.images = hotel.images
  }

  public get getId(): number {
    return this.id
  }

  public get getName(): string {
    return this.name
  }

  public get getDescription(): string {
    return this.description
  }

  public get getStarRating(): string {
    return 'â˜…'.repeat(Number(this.stars))
  }

  public get getLocation(): string {
    return `${this.address.fullAddress}, ${this.address.city}, ${this.address.state}, ${this.address.zipCode || 'N/A'}`
  }

  public get isFavorite(): boolean {
    return this.favorite
  }

  public get hasAmenities(): boolean {
    return this.amenities !== null && this.amenities.length > 0
  }

  public get getPrice(): string {
    return formatCurrency(this.price)
  }

  public get getHasBreakfast(): boolean {
    return this.hasBreakFast
  }

  public get getStars(): string {
    return this.stars
  }

  public get getThumb(): string {
    return this.thumb
  }

  public get getImages(): string[] {
    return this.images
  }

  public get isRefundable(): boolean {
    return this.hasRefundableRoom && !this.nonRefundable
  }

  public get isAgreement(): boolean {
    return this.hasAgreement
  }

  public get getRoomsQuantity(): number {
    return this.roomsQuantity
  }

  public get getAmenities(): Amenity[] | null {
    return this.amenities
  }

  public get getPricePerNight(): string {
    return formatCurrency(Number(this.price / this.roomsQuantity))
  }
}
