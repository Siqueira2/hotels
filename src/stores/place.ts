import { defineStore } from 'pinia'
import { ref } from 'vue'

import type { Place } from 'src/interfaces'

import { fetchPlaces as fetchPlacesApi } from 'src/api'

export const usePlaceStore = defineStore('place', () => {
  const places = ref<Place[]>([])

  const fetchPlaces = async () => {
    try {
      const response = await fetchPlacesApi()
      places.value = response
    } catch (error) {
      return error
    }
  }

  return {
    places,
    fetchPlaces,
  }
})
