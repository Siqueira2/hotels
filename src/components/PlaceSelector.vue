<template>
  <div>
    <q-card flat>
      <h5 class="text-grey-2 q-px-md q-py-sm q-ma-none">
        {{ $t('hotels.title') }}
      </h5>

      <q-separator />

      <q-card-section>
        <q-form @submit="onSubmit">
          <p class="text-weight-semibold q-mb-sm text-grey-6">
            {{ $t('hotels.search.label') }}

            <span class="text-red-5 text-weight-bold">*</span>
          </p>

          <q-select
            v-model="model"
            :options="filteredPlaces"
            dense
            outlined
            use-input
            hide-selected
            hide-dropdown-icon
            fill-input
            class="full-width row"
            :disable="loading"
            :loading
            input-debounce="0"
            :placeholder="$t('hotels.search.placeholder')"
            @filter="filterPlaces"
            @update:model-value="setModel"
          >
            <template #no-option>
              <div class="q-pa-md">
                {{ $t('hotels.search.no-results') }}
              </div>
            </template>
          </q-select>

          <div class="row justify-end q-mt-md">
            <q-btn
              :label="$t('hotels.search.action')"
              color="primary"
              type="submit"
              rounded
              unelevated
            />
          </div>
        </q-form>
      </q-card-section>
    </q-card>
  </div>
</template>

<script setup lang="ts">
import { ref, onMounted, computed } from 'vue'
import { storeToRefs } from 'pinia'
import { useRouter, useRoute } from 'vue-router'

import type { Place } from 'src/models/place'

import { usePlaceStore } from 'src/stores/place'

type FormattedPlace = Place & { label: string; value: number }

const router = useRouter()
const route = useRoute()
const placeStore = usePlaceStore()
const { places } = storeToRefs(placeStore)
const { fetchPlaces } = placeStore

const model = ref<string | null>(null)
const loading = ref<boolean>(false)

onMounted(async () => {
  loading.value = true
  await fetchPlaces()
  loading.value = false
})

const formattedPlaces = computed(() =>
  places.value.map((place) => ({
    label: place.formattedName,
    value: place.getPlaceId,
    ...place,
  })),
)

const filteredPlaces = ref<FormattedPlace[]>([])
const setFilteredPlaces = () => {
  filteredPlaces.value = formattedPlaces.value
}

const setModel = (place: FormattedPlace) => {
  model.value = place.label
}

const filterPlaces = (search: string, update: (fn: () => void) => void) => {
  setFilteredPlaces()
  update(() => {
    const searchText = search.toLowerCase()
    filteredPlaces.value = formattedPlaces.value.filter((place) =>
      place.label.toLowerCase().includes(searchText),
    )
  })
}

const onSubmit = async () => {
  if (!model.value) return

  const selectedPlace = formattedPlaces.value.find((place) => place.label === model.value)
  if (!selectedPlace) return

  await router.push({
    query: {
      ...route.query,
      name: selectedPlace.label,
    },
  })
}
</script>
