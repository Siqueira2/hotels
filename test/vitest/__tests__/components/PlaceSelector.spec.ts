import { describe, it, expect, beforeEach, vi } from 'vitest'
import type { VueWrapper } from '@vue/test-utils'
import { mount } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'
import type { ComponentPublicInstance } from 'vue'

import { usePlaceStore } from 'src/stores/place'
import { Place as PlaceModel } from 'src/models'
import type { Place as PlaceType } from 'src/interfaces'

import PlaceSelector from 'src/components/PlaceSelector.vue'

import PLACE from 'src/data/place.json'

vi.mock('vue-router', () => {
  return {
    useRoute: () => ({ query: {} }),
    useRouter: () => ({ push: vi.fn() }),
  }
})

const placeMock = new PlaceModel(PLACE[0] as PlaceType)

const normalizeTestIds = (html: string) =>
  html.replace(/(?:id|aria-controls|for)="f_[a-z0-9-]+(?:_lb)?"/g, (match) =>
    match.includes('aria-controls')
      ? 'aria-controls="f_test-id_lb"'
      : `${match.split('=')[0]}="f_test-id"`,
  )

const createWrapper = () =>
  mount(PlaceSelector, {
    global: { plugins: [createTestingPinia({ createSpy: vi.fn })] },
  }) as VueWrapper<ComponentPublicInstance<typeof PlaceSelector>>

describe('PlaceSelector', () => {
  let wrapper: VueWrapper<ComponentPublicInstance<typeof PlaceSelector>>
  let store: ReturnType<typeof usePlaceStore>

  beforeEach(() => {
    wrapper = createWrapper()
    store = usePlaceStore()
    store.places = [placeMock]
  })

  it('should match snapshot', () => {
    const wrapperHtml = normalizeTestIds(wrapper.html())
    expect(wrapperHtml).toMatchSnapshot()
  })

  it('should call fetchPlaces when mounted', () => {
    expect(store.fetchPlaces).toHaveBeenCalled()
  })

  it('should filter places based on user input', async () => {
    const input = wrapper.find('input')
    await input.setValue('S達o Paulo')

    const filteredPlaces = store.places.filter((place) => place.formattedName.includes('S達o Paulo'))
    expect(wrapper.vm.filteredPlaces).toEqual(filteredPlaces)
  })

  it('should enable the button when a place is selected', async () => {
    const input = wrapper.find('input')
    await input.setValue('S達o Paulo')

    const select = wrapper.findComponent({ name: 'QSelect' })
    const options = select.vm.$props.options

    expect(options).toBeDefined()

    await select.setValue(options[0]?.value)

    const button = wrapper.find('button[type="submit"]')
    expect(button.attributes('disabled')).toBeUndefined()
  })

  it('should not show no-option message when places are found', async () => {
    const input = wrapper.find('input')
    await input.setValue('S達o Paulo')

    const noOptionMessage = wrapper.find('.q-pa-md')
    expect(noOptionMessage.exists()).toBe(false)
  })
})
