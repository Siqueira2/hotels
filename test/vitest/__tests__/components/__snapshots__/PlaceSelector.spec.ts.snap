// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlaceSelector > should match snapshot 1`] = `
"<div>
  <div class=\\"q-card q-card--flat no-shadow\\">
    <h5 class=\\"text-grey-2 q-px-md q-py-sm q-ma-none\\">hotels.title</h5>
    <hr class=\\"q-separator q-separator--horizontal\\" aria-orientation=\\"horizontal\\">
    <div class=\\"q-card__section q-card__section--vert\\">
      <form class=\\"q-form\\">
        <p class=\\"text-weight-semibold q-mb-sm text-grey-6\\">hotels.search.label <span class=\\"text-red-5 text-weight-bold\\">*</span></p><label for=\\"f_test-id\\" class=\\"q-field row no-wrap items-start q-field--outlined q-select q-field--auto-height q-select--with-input q-select--without-chips q-select--single q-field--dense full-width row\\">
          <!---->
          <div class=\\"q-field__inner relative-position col self-stretch\\">
            <div class=\\"q-field__control relative-position row no-wrap\\" tabindex=\\"-1\\">
              <div class=\\"q-field__control-container col relative-position row no-wrap q-anchor--skip\\">
                <div class=\\"q-field__native row items-center\\"><input class=\\"q-field__input q-placeholder col\\" type=\\"search\\" tabindex=\\"0\\" role=\\"combobox\\" aria-readonly=\\"false\\" aria-autocomplete=\\"list\\" aria-expanded=\\"false\\" aria-controls=\\"f_test-id_lb\\" placeholder=\\"hotels.search.placeholder\\" id=\\"f_test-id\\" value=\\"\\"></div>
                <!---->
              </div>
              <!---->
              <!---->
            </div>
            <!---->
          </div>
          <!---->
        </label>
        <div class=\\"row justify-end q-mt-md\\"><button class=\\"q-btn q-btn-item non-selectable no-outline q-btn--unelevated q-btn--rectangle q-btn--rounded bg-primary text-white q-btn--actionable q-focusable q-hoverable\\" tabindex=\\"0\\" type=\\"submit\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><span class=\\"block\\">Buscar</span></span></button></div>
      </form>
    </div>
  </div>
</div>"
`;
